lib/internal/fs/promises.js

```diff
@@ -792,48 +794,72 @@ async function readdirRecursive(originalPath, options) {
       ),
     ],
   ];
-
+  function filter(absolutePath) {
+    const directory = absolutePath.slice(0, absolutePath.lastIndexOf(sep));
+    return absolutePath === originalPath ||
+      typeof options.recursiveFilter !== 'function' ||
+      (typeof options.recursiveFilter === 'function' && options.recursiveFilter({
+        depth: absolutePath.match(lengthReg).length - baseDirLength,
+        itemName: absolutePath.slice(absolutePath.lastIndexOf(sep) + 1),
+        itemPath: absolutePath,
+        parentName: directory.slice(directory.lastIndexOf(sep) + 1),
+        parentPath: directory,
+      }));
+  }
+  const sep = pathModule.sep;
+  const lengthReg = (sep === '\\') ?
+    new RegExp('\\\\', 'g') :
+    new RegExp('/', 'g');
+  const baseDirLength = originalPath.match(lengthReg).length;
```

Comment:
Wouldn't it be better to start with `let depth = 0;` and use `[path, readdir, depth + 1]` as queue item?

Reply:
In this approach the base path passed into the recursive function is depth 0.  It's immediate children are depth 1, and the immediate children of the immediate child directories are depth 2, and so forth.  It is counting the number of steps removed from the specified address.

Reply:
Sorry, I should've worded it better.
I meant something like:
```mjs
// base directory inits with depth = 0
const queue = [
  [
    originalPath,
    await binding.readdir(...),
    0, // initial depth
  ],
];
```
```mjs
// processing each element from queue
const { 0: path, 1: readdir, 2: depth } = ArrayPrototypePop(queue);
for (const ent of readdir) {
// ...
  if (stat === 1) {
    ArrayPrototypePush(queue, [
      direntPath,
      await binding.readdir(...),
      depth + 1, // subdirectories will have 1, subsubdirectories will have 2, etc.
    ]);
// ...
```
This would achieve the same thing with better performance and robustness, avoiding parsing path every time, counting separators, and memorizing `baseDirLength`.